### ISSUE TITLE
Remove and replace QueryableOrderedSet

### ISSUE THREAD

**Issue by st-pasha** [flame-engine/flame#976]
Our current implementation of the `ComponentSet` is built upon the class `QueryableOrderedSet`, which in turn extends `OrderedSet`, which is built on top of a `SplayTreeSet`.

The choice of the container structure used throughout the game engine is obviously a very important one. Such choice should first and foremost be based on the efficiency consideration, as well the set of features that various implementations support. Thus, I'd like to compare the current implementation versus a more simple one based on a plain list of Components.

1. The most common pattern of access to the components is iteration: we iterate through all of the components at least twice every tick: once for the update()s, and then once again for render()ing. It is dead simple to iterate through a plain list. Though one can argue that `SplayTreeSet` also provides a suitable iterator, so this is simple too. Still, in terms of the number of CPU instructions, iterating over a list is ought to be simpler than iterating over a tree, so it's probably a small win for the List approach.

2. The second most common pattern of access is an iteration in reverse order, which we perform for every pointer event. Many games rely on drag events to control the main character (for example via joystick), and these events fire every game tick. Thus, on every game tick, we need to iterate through the game components in reverse order. Once again, iterating in reverse over a list is very simple. For the current `QueryableOrderedSet`, it's not: in order to iterate over the tree backwards, we convert it to a list and then reverse that list. Needless to say, serializing the entire game tree into a list on every game tick is rather expensive and should be avoided. Thus, a clear win for the List approach.

3. Extendability. If `QueryableOrderedSet` was implemented in-house, it would be easy to extend it with new functionality. For example, it would be possible to add functionality to modify the set while iterating over it; or to replace an existing element with a new one. It is very challenging to do the same with the existing implementation since many of its internal functions are private and thus cannot be overridden. Again, this is a win for the List approach.

4. But what about ordering components by their priority? Well, first I'd like to point out that the majority of games do not bother with setting priorities for the components. At most, they are interested in adding new elements on top of existing ones or move one of the existing elements on top of the others. Both of these operations are trivial to implement with a simple List. If a game needs to add new components at runtime somewhere in the middle of the other components, then the sanest approach would be to manually introduce several distinct layers within the game and then add new components to the appropriate layer. If, for whatever reason, this approach cannot be followed, and the user wants to add a component with a non-zero priority inside a list of other components -- then yes, the SplayTree approach would be more efficient here. When adding multiple such components at once, the List approach would simply add these components at the end and then sort the list. The SplayTree approach in this case serializes everything into a list and then re-creates the tree from the list. Thus, in this case, both approaches are roughly equivalent. Overall, SplayTree probably wins this round, albeit with a very minor advantage.

5. What about the "Queryable" part? Clearly, we can implement the equivalent functionality for the List approach -- we just need to keep a map of sublists corresponding to each query. In fact, we could improve upon at least one aspect of the current mechanism: there should be no need to "register" the types of queries beforehand. Making the same improvement within the `QueryableOrderedSet` is hard because it's a third-party code and we cannot override private methods there. So, again a small win for the List approach.

6. By the same argument, fixing bugs within third-party `QueryableOrderedSet` is difficult. For example, currently the `query<T>` subsets are not ordered: i.e. their order does not correspond to the order of the elements in the overall ordered set. This is not documented anywhere, and may not be expected, so a potential bug. Another problem is that `rebalanceWhere()` will create duplicate elements within the queries. It's unclear how severe this problem is since the method is probably rarely used -- but still, we'd have a much better chance to fix issues like this if we had an in-house implementation of the QueryableOrderedSet. Thus, once again, a modest win for the List approach.

Overall, I feel like this is a clear win for the List approach here.

**Comment by QiXi** [comment]
And more
slowly:
```dart
for (var item in list) {
...
}
```
quickly:
```dart
final length = list.length;
for (var i = 0; i < length; i++) {
...
}
```

**Comment by erickzanardo** [comment]
Ordering and priority is a big deal in my opinion, this is a feature that has been requested and improved throughout Flame history. So losing this is a no go for me.

I feel that the major argument here is that it is hard to improve this because ordered set is a third party package. Which isn't entirely true, we maintain that package ourselves, it is just in a separated package because organizational wise, it makes sense.

Ordered set provides many features that helps the flame code base, specially the query ones, so I don't think we should stop using it, but rather improve it if there are any bug or improvements to be made there.

**Comment by luanpotter** [comment]
Yeah, it's defitenily not a third-party! we wrote it :) I agree with erick, seems most concerns are related to that which is not true.

For the missing things on ordered set, we can def improve it!

> reverse iterator
we should add support for that, it should be fairly simple to impl

> there should be no need to "register" the types of queries beforehand
I don't see how that could be accomplished w/o paying a performance hit on the first usage

> query<T> subsets are not ordered
that was never a concern for me, but we can implement that if desired. I wonder if it would add any extra cost though, if most cases we don't care when we do might be cheaper to just sort the sublist

**Comment by luanpotter** [comment]
@QiXi I didn't understand your comment

**Comment by erickzanardo** [comment]
Closing this as we want to keep the ordered set, please any issues of improvements that can be done on ordered set, open issues or even PRs on its repository.

Thanks!

**Comment by QiXi** [comment]
> @QiXi I didn't understand your comment

If you want to switch to List and want to measure performance, keep in mind that using Iterable is slower than index access.
or for example using .forEach () is not good either. Game engines are not a place where performance issues can be neglected.

**Comment by QiXi** [comment]
IMHO: the final word for the choice of the architectural approach to be behind the performance tests

**Comment by st-pasha** [comment]
> Ordering and priority is a big deal, in my opinion, this is a feature that has been requested and improved throughout Flame history. So losing this is a no-go for me.

I'm not suggesting losing these properties. After all, a List is perfectly ordered. And we will keep the priorities as well: inserting an element into a sorted list is a fairly quick operation. The question here is the relative frequency of various operations. How often do we need to add new components to the game? Well, maybe there are many enemies that constantly appear, plus some special effects going on -- still, even if we add things 6 times per second, that operation will occur only once every 10 ticks.

Ok, but when we do add new components: which approach will be faster? This will depend on where you add those components. If they are added at the end, using the default priority level, then using a list will be faster than using a SplayTree. After all, a SplayTree is highly volatile and rebalances itself on every operation (in fact, SplayTree is the worst implementation of a BST for our use case: a RedBlackTree, or an AvlTree would have been much better). The advantage of the tree approach only appears when you add a component in the middle of the existing component list, and when that list is fairly large.

I would argue that such use cases are very rare, and perhaps even indicative of bad design. We have moved away from the general paradigm of "add all your components as children of the BaseGame object". Now we prefer for the components to be organized into a component tree by their logic: a background layer, a foreground layer, a layer for enemies, etc. With this design, each component's children list is fairly small, and the priorities within that list are fairly uniform.

Basically, what we have right now is a tree of components, but every component carries a mini-tree inside to organize its children, and this tree-of-trees structure seems overly complicated and unnecessary to me.

Of course, these considerations are somewhat theoretical. It would be more helpful to talk about a specific game example -- @erickzanardo could you suggest an example of a game where you believe it would be beneficial to have the component set as a tree?

> Yeah, it's defitenily not a third-party! we wrote it :)

Did not know that! My Android Studio shows it as third-party, so I assumed it was.

**Comment by erickzanardo** [comment]
> Of course, these considerations are somewhat theoretical. It would be more helpful to talk about a specific game example -- @erickzanardo could you suggest an example of a game where you believe it would be beneficial to have the component set as a tree?

I must confess that I don't enough knowledge of tree algorithms to think on an example. But I think this isn't the point.

The point here is that we want to keep using the ordered set, which by my understanding can have different algorithm providing its features under the hood, be it a tree or not. Ideally ordered set could maybe provide some options of implements so the user could choose the one that fits better?

Anyhow, if there is a different set algorithm that will bring significant perfomance to flame, I am totally on boarding on adding it to ordered_set and use it on flame! 

But this discussion should probably be on a different place than this issue

**Comment by st-pasha** [comment]
@erickzanardo If I understand you correctly, you're not opposed to the idea _per se_, as long as the set of features doesn't diminish, and performance doesn't degrade, and there are at least some clear wins with the new approach?
Also, what do you think would be a better venue for having this discussion?

**Comment by erickzanardo** [comment]
@st-pasha the only think I was against, was to stop using `ordered_set` and placing a simple list instead 😉 

We probably could use github discussions for discussing this other ideas.
